blog/terraform-post.md

```diff
@@ -0,0 +1,67 @@
+
+Let's take a look at a pseudocode version of the Terraform lifecycle.
+
+```
+userIntention := GetTerraformPlan()
+currentState := ReadResource(userIntention)
+
+if currentState != null:
+	if user wishes to delete:
+		 DeleteResource(currentState)
+	else if currentState != userIntention:
+		 UpdateResource(userIntention)
+else:
+	CreateResource(userIntention)
+```
+
+Two different sets of problems emerge and the AEPs are well-suited for solving both of them.
+
+The first set of problems revolves around the grouping and calling of APIs. There's a variety of tools in the market that will call an API given a set of parameters and an OpenAPI description (or even just a URL).Terraform doesn't expose individual APIs, but instead exposes a resource schema. This raises a variety of questions when developing a Terraform provider:
+
+* What are my resources?  
+* How do I know which APIs to call for a given resource?  
+* Is there 1 API that can be called for each CRUD action, or will I need to chain together API calls to achieve my result?  
+* How do I derive my resource schema from 4 distinct CRUD APIs? Is it simply the combination of all the body parameters from the 4 APIs?
+
+AEP APIs are resource-oriented by default, so just following the AEPs can address all of the problems above. The Terraform provider can make assumptions about which APIs should be called and how they relate to each other.
+
+# What about API responses?
+
+The second set of problems comes from reading API responses. In the pseudocode above, the Read response needs to be parsed and checked against the Terraform plan. This means that the Terraform provider needs to understand the relationship between the API response (beyond simple error checking) and the user's intention. More questions emerge:
+
+* Does the response of the Read API match the resource schema? What changes need to be made?  
+* Are there multiple ways to express a value in the cloud? If so, how can I be sure that two semantically equivalent values do not result in a diff against the Terraform plan?  
+* If a boolean value does not appear in the JSON response, does that mean it's false? (this can be extended to all types, not just booleans\!)
+
+Many of these questions come down to the preference of the API developer. In fact, none of them can be defined by an API schema\! The AEPs encode these differences in our specification, so that tools like our Terraform provider can make assumptions about how to parse the responses of our APIs. 
+
+# What comes next?
+
+While the Terraform provider lifecycle appears simple at first glance, the details contain a deceptive amount of detail. This is detail that can only be learned through trial-and-error, which is difficult when APIs are often only designed once.
+
+By building out the Terraform provider, we're even more convinced than ever that the AEP specification can prove the thesis that client tooling can be more easily built on top of consistent APIs.
```

Comment:
WDYT about adding more detail on the actual Terraform provider?  This ending felt kind of abrupt to me.  You wrote above that "We’d like to tell you about the first version of a fully auto-generated Terraform provider that can be used with AEP-compliant APIs." but then this doesn't ever exactly happen.  The post talks about why an AEP-based Terraform provider generator could work and be valuable, but it doesn't really talk about the fact that you've implemented one or what it looks like!

Reply:
Check out the new paragraph I added.